### PR TITLE
publish cart vel magnitudes for frames of chain

### PR DESCRIPTION
**Pull request by fmessmer**
ref https://github.com/mojin-robotics/cob4/issues/2016

uses http://docs.ros.org/en/jade/api/orocos_kdl/html/classKDL_1_1ChainFkSolverVel__recursive.html

### CODE REVIEW

**Comment by fmessmer** [comment]
goes best with https://github.com/mojin-robotics/mojin_robots/pull/408

```
roslaunch cob_twist_controller debug_evaluate_jointstates.launch component:=arm chain_base_link:=arm_base_link_inertia chain_tip_link:=gripper_ee_fixed_link
```

**Comment by fmessmer** [comment]
Node has been renamed to 
```
roslaunch cob_twist_controller debug_fk_vel_recursive.launch component:=arm chain_base_link:=arm_base_link chain_tip_link:=gripper_ee_fixed_link
```

File: cob_twist_controller/src/debug/debug_fk_vel_recursive_node.cpp [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        KDL::JntArrayVel jntArrayVel = KDL::JntArrayVel(q, q_dot);
+        v_FrameVel.resize(chain_.getNrOfSegments());
+        if (p_fksolver_vel_->JntToCart(jntArrayVel, v_FrameVel, chain_.getNrOfSegments()) >= 0)
+        {
+            //last entry is twist of tip_link
+            geometry_msgs::Twist twist_msg;
+            tf::twistKDLToMsg(v_FrameVel.back().GetTwist(), twist_msg);
+            twist_current_pub_.publish(twist_msg);
+
+            //recursively calculate FrameVel magnitudes
```
Comment (LoyVanBeek) [inline cob_twist_controller/src/debug/debug_fk_vel_recursive_node.cpp:155]: I don't see what is recursive here. Code just iterates over the joints right?

Or do I understand wrong?

Comment (LoyVanBeek) [reply cob_twist_controller/src/debug/debug_fk_vel_recursive_node.cpp:155]: Ah, the FK solver works recursively, it's a https://docs.ros.org/en/indigo/api/orocos_kdl/html/classKDL_1_1ChainFkSolverVel__recursive.html

**Review by LoyVanBeek — COMMENTED** [review]
I think the code makes sense. 

To summarize and check my understanding:
When we receive a `JointTrajectoryControllerState` message, we set up arrays of their positions and velocities, pass it off to a KDL forward kinematics solver (instantiated from the URDF robot description) that does the core calculation (treating that as a black box). The solver calculates Cartesian velocities from the joint positions and velocities, which are then put back into a Twist. 

Before we use this for calculating derivatives: currently the Norm of the velocity is taken, so it cannot go negative. That might have an effect on the correctness of the derivatives of course.

File: cob_twist_controller/src/debug/debug_fk_vel_recursive_node.cpp [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        KDL::JntArrayVel jntArrayVel_actual = KDL::JntArrayVel(q_actual, q_actual_dot);
+        v_FrameVel_actual.resize(chain_.getNrOfSegments());
+        if (p_fksolver_vel_->JntToCart(jntArrayVel_actual, v_FrameVel_actual, chain_.getNrOfSegments()) >= 0)
+        {
+            sensor_msgs::JointState magnitude_actual_msg;
+            magnitude_actual_msg.header.stamp = msg->header.stamp;
+            for (unsigned int i = 0; i < chain_.getNrOfSegments(); i++)
+            {
+                magnitude_actual_msg.name.push_back(chain_.getSegment(i).getName());
+                magnitude_actual_msg.velocity.push_back(v_FrameVel_actual[i].GetTwist().vel.Norm());
```
Comment (LoyVanBeek) [inline cob_twist_controller/src/debug/debug_fk_vel_recursive_node.cpp:237]: Not sure if taking the Norm is valid when we want to use derive the acceleration from these later on. 
Why take the ||norm||? Having a negative velocity should be fine, right?

Comment (fmessmer) [reply cob_twist_controller/src/debug/debug_fk_vel_recursive_node.cpp:237]: derivatives are not of interest for the time being....magnitudes are fine for the safety assessment

Comment (LoyVanBeek) [reply cob_twist_controller/src/debug/debug_fk_vel_recursive_node.cpp:237]: Allright then. But still to keep in mind maybe.

**Review by LoyVanBeek — DISMISSED** [review]

**Comment by fmessmer** [comment]
successfully used for the analysis in https://github.com/mojin-robotics/cob4/issues/2016

**Pull request merged**